### ISSUE TITLE
BeginScope forgets stuff when returning from the method

### ISSUE THREAD

**Issue by mariusGundersen** [NLog/NLog#3547]
**NLog version**: 4.5.0

**Platform**:  .NET Core 2.1

```csharp
        [HttpGet]
        public async Task<IActionResult> Test(int id)
        {
            var item = await GetItem(id);

            _injectedLogger.LogInformation("Doing something with the item");

            return Ok();
        }

        private async Task<object> GetItem(int id)
        {
            _injectedLogger.BeginScope(new Dictionary<string, object>{["id"] = id});

            // pretend we are getting it from the database, and it takes a long time
            await Task.Delay(100);

            _injectedLogger.LogInformation($"Fetched an object from the database with id={id}");

            // pretend this is the thing we got from the database
            return new { id };
        }
```

In the above sample I call `BeginScope` inside the `GetItem` method, setting up the logger to include the id for the rest of the request processing. But, the id is only included in the logged message inside `GetItem`, in the log of the message "Doing something wit hthe item" it is missing. It seems like the value is forgotten when the async method `GetItem` returns.

**Comment by snakefoot** [comment]
You are doing the following:

- Test()
   - Starts new GetItem()-task-scope
     - Inserts scope-item "id" for this scope
     - awaits delay
     - LogInformation("fetched - Captured id")
   - Returns from GetItem()-task-scope
     - LogInformation("Do something- No capture of scope-item from other task")

When creating a new task-scope, then any context modifications you perform within the new task-scope will not affect the calling task.

This is fully intentionally to avoid unwanted side-effects. Other example just for fun:

- Test()
 - Inserts scope-item "method" = "Test"
 - Starts new GetItem()-task-scope
    - Inserts scope-item "method" = "GetItem"
    - LogInformation("fetched - Captured method = GetItem")
 - Returns from GetItem()-task-scope
   - LogInformation("Do something- Captured method = Test")

If the BeginScope behave like you wanted then the last LogInformation would have done this:

     - LogInformation("Do something - Captured method = GetItem")

You want task-scope-modification to flow to new tasks, but you don't want task-scope-information from new tasks to return back to the calling task. Same misunderstanding here #3301

**Comment by snakefoot** [comment]
The correct way to implement the logic would be this:

```c#
        [HttpGet]
        public async Task<IActionResult> Test(int id)
        {
           using (_injectedLogger.BeginScope(new KeyValuePair<string, object>("id", id))
           {
                var item = await GetItem(id);
               _injectedLogger.LogInformation("Doing something with the item");
               return Ok();
            }
        }
```

**Comment by mariusGundersen** [comment]
Thanks for the explanation. I see how it works, but I disagree with why that is the most useful way it could work. For example in my scenario I call the `GetItem(id)` method from many different places and want the `id` to be a part of all logging after that point. Having to repeat the `BeginScope` everywhere `GetItem` is called is verbose and it's easy to forget it. In my actual scenario I have several other properties of the item fetched from the database that I also want to attach to the logging context, so I have to repeat about 6 lines of `BeginScope` everywhere I call `GetItem`. It would clearly be more useful if I could do it only once, inside `GetItem`. If I didn't want the context to apply to the calling scope as well then I could use the `using(){  }` form, where it would be disposed at the end of the block. That way I can choose if I want it to escape the called scope or not.

Is there no way to achive this?

**Comment by snakefoot** [comment]
> Is there no way to achive this?

This sounds like a question for StackOverflow. I'm not that skilled with Threading-Tasks (usually work with dedicated threads pinned to cores to optimize speed and reduce allocation).

Maybe this will solve the issue;

```
Task<object> GetItem(int id)
{
    injectedLogger.BeginScope(new Dictionary<string, object>{["id"] = id});
    return GetItemAsync(id);
} 

private async Task<object> GetItemAsync(int id)
{
   await Task.Delay(100);
   return new { id };
}
```

But the NLog implementation of BeginScope is very close to what you get when using pure  Microsoft-Extension-Logging (MEL) Logging Provider. So any solution suggested by Microsoft Developers will also work with NLog as logging-provider in MEL.

**Comment by 304NotModified** [comment]
I assume your question has been answered, if not, please let us know!
